woocommerce/woocommerce-android
# Update automation that freezes strings for translation

### Description

After the prep work from #6741 and #6703, this PR updates the automation to freeze the strings for translation.

Notably it:

- Uses the new `source_id` parameter to differentiate the generated strings
- Uses the Fastlane APIs to interact with Git
- Uses a structure consistent with WordPress Android

### Testing instructions

To test this, I created a new branch from this one, hacked the `complete_code_freeze` lane to work outside of a release branch, and used it to verify the process succeeds. You can see it, including the commit with the new frozen strings, in #6800, but for reference here's the lane output

```
➜ bf complete_code_freeze

[✔] 🚀
[14:46:14]: ------------------------------
[14:46:14]: --- Step: default_platform ---
[14:46:14]: ------------------------------
[14:46:14]: Driving the lane 'android complete_code_freeze' 🚀
[14:46:14]: -------------------
[14:46:14]: --- Step: is_ci ---
[14:46:14]: -------------------
[14:46:14]: ---------------------------------------
[14:46:14]: --- Step: check_for_toolkit_updates ---
[14:46:14]: ---------------------------------------
[14:46:14]: Currently using release toolkit version 4.2.0.
[14:46:14]: Checking for updates now...
[14:46:16]: Your release toolkit is up-to-date! ✅
[14:46:16]: -------------------------------------------------------
[14:46:16]: --- Step: Switch to android localize_libraries lane ---
[14:46:16]: -------------------------------------------------------
[14:46:16]: Cruising over to lane 'android localize_libraries' 🚖
[14:46:16]: ----------------------------------------------
[14:46:16]: --- Step: android_download_file_by_version ---
[14:46:16]: ----------------------------------------------
[14:46:16]: Downloading WordPressLoginFlow/src/main/res/values/strings.xml from wordpress-mobile/WordPress-Login-Flow-Android at version 0.13.0 to /var/folders/29/wn6913pn73gf2l7j5n4n56mw0000gn/T
[14:46:17]: Logged in as: Gio Lodi
[14:46:17]: Strings.xml file for Login Library downloaded to /var/folders/29/wn6913pn73gf2l7j5n4n56mw0000gn/T/strings.xml.
[14:46:17]: ------------------------------
[14:46:17]: --- Step: an_localize_libs ---
[14:46:17]: ------------------------------
[14:46:17]: Merging Login Library strings into WooCommerce/src/main/res/values/strings.xml
[14:46:17]:  - Skipping enter_email_wordpress_com string
[14:46:17]:  - Skipping enter_email_for_site string
[14:46:17]:  - Skipping enter_site_address string
[14:46:17]:  - Skipping default_web_client_id string
[14:46:17]:  - Skipping login_notification_channel_id string
[14:46:17]: Done (0 added, 0 updated, 117 untouched, 5 skipped).
[14:46:17]: ----------------------------------------------
[14:46:17]: --- Step: android_download_file_by_version ---
[14:46:17]: ----------------------------------------------
[14:46:17]: Downloading library/src/main/res/values/strings.xml from Automattic/about-automattic-android at version 0.0.4 to /var/folders/29/wn6913pn73gf2l7j5n4n56mw0000gn/T
[14:46:18]: Strings.xml file for About Library downloaded to /var/folders/29/wn6913pn73gf2l7j5n4n56mw0000gn/T/strings.xml.
[14:46:18]: ------------------------------
[14:46:18]: --- Step: an_localize_libs ---
[14:46:18]: ------------------------------
[14:46:18]: Merging About Library strings into WooCommerce/src/main/res/values/strings.xml
[14:46:19]: Done (0 added, 0 updated, 27 untouched, 0 skipped).
[14:46:19]: ------------------------
[14:46:19]: --- Step: git_commit ---
[14:46:19]: ------------------------
[14:46:19]: $ git status WooCommerce/src/main/res/values/strings.xml --porcelain
[14:46:19]: Nothing to commit, working tree clean ✅.
[14:46:19]: Cruising back to lane 'android complete_code_freeze' 🚘
[14:46:19]: --------------------------------------------------------------------------
[14:46:19]: --- Step: Switch to android update_frozen_strings_for_translation lane ---
[14:46:19]: --------------------------------------------------------------------------
[14:46:19]: Cruising over to lane 'android update_frozen_strings_for_translation' 🚖
[14:46:19]: ------------------------
[14:46:19]: --- Step: git_commit ---
[14:46:19]: ------------------------
[14:46:19]: $ git status ./fastlane/resources/values/strings.xml --porcelain
[14:46:19]: ▸ M fastlane/resources/values/strings.xml
[14:46:19]: $ git commit -m Freeze\ strings\ for\ translation ./fastlane/resources/values/strings.xml
[14:46:19]: ▸ [test/6742 b0a0fede9b] Freeze strings for translation
[14:46:19]: ▸ 1 file changed, 11 insertions(+), 1 deletion(-)
[14:46:19]: Successfully committed "["./fastlane/resources/values/strings.xml"]" 💾.
[14:46:19]: Cruising back to lane 'android complete_code_freeze' 🚘
[14:46:19]: -------------------------------------
[14:46:19]: --- Step: ensure_git_status_clean ---
[14:46:19]: -------------------------------------
[14:46:19]: $ git status --porcelain
[14:46:19]: Git status is clean, all good! 💪
[14:46:19]: --------------------------------
[14:46:19]: --- Step: push_to_git_remote ---
[14:46:19]: --------------------------------
[14:46:19]: $ git push origin test/6742:test/6742 --tags
[14:46:28]: ▸ remote:
[14:46:28]: ▸ remote: Create a pull request for 'test/6742' on GitHub by visiting:
[14:46:28]: ▸ remote:      https://github.com/woocommerce/woocommerce-android/pull/new/test/6742
[14:46:28]: ▸ remote:
[14:46:28]: ▸ remote: GitHub found 1 vulnerability on woocommerce/woocommerce-android's default branch (1 low). To find out more, visit:
[14:46:28]: ▸ remote:      https://github.com/woocommerce/woocommerce-android/security/dependabot/8
[14:46:28]: ▸ remote:
[14:46:28]: ▸ To github.com:woocommerce/woocommerce-android
[14:46:28]: ▸ * [new branch]            test/6742 -> test/6742
[14:46:28]: Successfully pushed to remote.

+------+----------------------------------------------------------+-------------+
|                               fastlane summary                                |
+------+----------------------------------------------------------+-------------+
| Step | Action                                                   | Time (in s) |
+------+----------------------------------------------------------+-------------+
| 1    | default_platform                                         | 0           |
| 2    | is_ci                                                    | 0           |
| 3    | check_for_toolkit_updates                                | 2           |
| 4    | Switch to android localize_libraries lane                | 0           |
| 5    | android_download_file_by_version                         | 0           |
| 6    | an_localize_libs                                         | 0           |
| 7    | android_download_file_by_version                         | 1           |
| 8    | an_localize_libs                                         | 0           |
| 9    | git_commit                                               | 0           |
| 10   | Switch to android update_frozen_strings_for_translation  | 0           |
|      | lane                                                     |             |
| 11   | git_commit                                               | 0           |
| 12   | ensure_git_status_clean                                  | 0           |
| 13   | push_to_git_remote                                       | 9           |
+------+----------------------------------------------------------+-------------+

[14:46:28]: fastlane.tools finished successfully 🎉
```

---
